### DISCUSSION TITLE
Filter repositories by the topic

### DISCUSSION THREAD

**Discussion by shayanmalinda** [community/community#69391] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

Is it possible to fetch the list of repositories of a organization filtered by a topic?
Although there's an endpoint to fetch the repos I couldn't filter by the topic?

**Comment by seyLu** [comment]
### Hi @shayanmalinda,

From the documentation, fetching repos from org also returns the topics associated with the repo:
![image](https://github.com/community/community/assets/98249191/73186eea-5b0d-439f-a538-44200165957a)

So we can write a script to filter the returned repos by topic: (for example, in python)
```py
import logging
import os
import sys
from typing import Any

import requests
from dotenv import load_dotenv
from requests import Response

load_dotenv()

if __name__ == "__main__":
    org: str = os.getenv("ORG", "<organization_name>")
    token: str = os.getenv("TOKEN", "<personal_access_token>")

    url: str = f"https://api.github.com/orgs/{org}/repos"
    headers: dict[str, str] = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
    }

    res: Response = requests.get(url, headers=headers)
    logging.info(f"Fetching repos from org {org}.")

    if not res.ok:
        logging.error(f"Status {res.status_code}. Failed to fetch list of repos.")
        sys.exit()

    repos: list[dict[str, Any]] = res.json()
    topics: list[str] = [
        "<insert_list_of_topics_here>",
    ]

    filtered_repos: list[dict[str, Any]] = []
    for topic in topics:
        filtered_repos.extend(
            [repo for repo in repos if topic in repo.get("topics", [])]
        )

    print([filtered_repo["name"] for filtered_repo in filtered_repos])
```

**Reply by shayanmalinda** [reply]
The problem with this is, GitHub API returns maximum of 100 repos per single request.
Hence, we can not use this approach.
Also I need to use the pagination feature in the API as well.

**Reply by seyLu** [reply, marked as answer]
And what is stopping us from applying pagination to the existing script?

```py
import logging
import os
import sys
from typing import Any

import requests
from dotenv import load_dotenv
from requests import Response

load_dotenv()
logging.getLogger().setLevel(logging.INFO)
logging.basicConfig(format="[%(levelname)s] %(name)s: %(message)s")

if __name__ == "__main__":
    org: str = os.getenv("ORG", "<organization_name>")
    token: str = os.getenv("TOKEN", "<personal_access_token>")

    topics: list[str] = [
        "<insert_list_of_topics_here>",
    ]

    url: str = f"https://api.github.com/orgs/{org}/repos"
    headers: dict[str, str] = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
    }

    page: int = 1
    per_page: int = 100

    logging.info(f"Fetching list of repos from org {org}.")

    repos: list[dict[str, Any]] = []
    filtered_repos: list[dict[str, Any]] = []
    while True:
        params: dict[str, int] = {"page": page, "per_page": per_page}
        res: Response = requests.get(
            url,
            headers=headers,
            params=params,
        )
        logging.info(f"Fetching page {page}.")

        if not res.ok:
            logging.error(f"Status {res.status_code}. Failed to fetch list of repos.")
            sys.exit()

        repos_on_page = res.json()
        if not repos_on_page:
            break

        repos.extend(repos_on_page)
        page += 1

    for topic in topics:
        filtered_repos.extend(
            [repo for repo in repos if topic in repo.get("topics", [])]
        )

    logging.info("Successfully filtered repos by topic.")
    print([filtered_repo["name"] for filtered_repo in filtered_repos])
```

**Reply by ghostinhershell** [reply]
Hey there @shayanmalinda, since you did mention needing to also use pagination with GitHub's API here, I thought I would jump in and provide this resource: [How To Use Pagination With GitHub's API](https://github.com/orgs/community/discussions/69826).

Hopefully this helps! I also ask that, if @seyLu's answers here have been helpful, that you mark his response as the answer so that other users who have a similar question can easily reference this Discussion in the future. If this doesn't answer your question, let us know so that we can further assist. 

Thanks!

**Reply by ghostinhershell** [reply]
Hey there @shayanmalinda, 

It's been a while since you've responded here. Since @seyLu's answer here did address the question that you had I'm going to go ahead and mark their response as the answer. Please feel free to reach back out if you have any additional questions. 

Cheers! 👋🏾

**Comment by shayanmalinda** [comment]
HI all,

Thanks for your suggestions and answers.

Recently, I found that there's a separate set of GitHub APIs which could use to filter the repositories by the topics.
Please refer to:
[`https://docs.github.com/en/search-github/searching-on-github/searching-for-repositories`](https://docs.github.com/en/search-github/searching-on-github/searching-for-repositories)

Again, appreciate your suggestions and anwers.

Thank you.
